src/cli/commands/force/lightning/local/setup.ts

```diff
@@ -33,18 +33,8 @@ export class Setup extends SfdxCommand implements HasRequirements {
     public static description = messages.getMessage('commandDescription');
 
     public static readonly flagsConfig: FlagsConfig = {
```

Comment:
In that case then there should be no flag, like `apilevel` or `platform`, to be consolidated else where, right? 

The root of the problem we faced was that `Setup` command being the base class to all other commands offering these flags and respective validation. So the solution was to port those into, `CommonUtils`(and in next PR to `CommonLineUtils`).

If we choose to have each command specify its own flags and not bring it from else where, like `CommonUtils`, I'm all for it. I am guessing how `SfdxCommand`(and to that extent oclif command) was originally designed was to be self-contained but we went off the hook with `Setup` command inheritance and sharing flags and validation logic.

Are we okay with the duplication of the code across the command form same flags?

Reply:
I changed the code so that `FlagConfig` type can be constructed so we can tell whether a flag is required or not. Please look at `createFlagConfig(...)`.

Reply:
Note, this makes us go back closer to the earlier idea of their being `FlagVerifierService` which serves as a central manager of flags and respective validation logic for different commands.

Reply:
So initially we did not have inheritance and none of the commands were extending from `Setup` (we only had 3 commands back then). As we were adding more commands, it was I who modified the code to have them all inherit from `Setup`, and I did that mainly to avoid duplicating the requirement checks. I wanted to spend most of my time implementing the command functionalities rather than doing a major refactor on how requirements are shared/executed. And since I was now inheriting from `Setup` I thought that perhaps I should share the flag/validation code too.

While it did help reduce code duplication in the short run, as you know in the long run that approach caused its own issues. Now that you have refactored execution of requirement checks, I think we can just do away with inheritance from `Setup` altogether.

I'm fine with defining flags in each command and the code duplication that it brings b/c it is very minimal, and it gives us the added flexibility of changing how the flags are defined per command.

We could use `createFlagConfig` that you created but I think we should allow for flag character to be passed too. A command might want to use `-a` for apiLevel and another command might want to use `-L` for apiLevel b/c it may want to use `-a` for something else. Besides, flags have other properties that we might use in the future and that might need customization. If we take this approach then in the future we would need to do yet another refactor.

Speaking of flag properties, I just noticed that `SfdxFlags` has a `validate` property where you can validate the value of the flag. We might be able to use that approach instead. Take a look at SfdxFlags.test.ts [in this commit](https://github.com/forcedotcom/cli-packages/commit/7d75bdd9b8ac36ef74b5c134c54c0b39b52c6fc9) as an example (there is also a way to set a function for `validate` instead of RegExp).

So perhaps we can simplify things by only defining 2 functions (for validating API Level and Platform) in CommandLineUtils to set as the `validate` functions. Then each command can define its own flags and can reference these 2 functions or provide its own functions!

Reply:
I tried to allow passing in a single alphabet character to create a flag. However, `oclif` wouldn't allow me that because that character needs to be a type of `AlphabetUppercase` or `AlphabetLowercase` which are defined in `@oclif/parser` 😞  I could have duplicated those types but with the likelihood of "api level" going away and also us eventually creating a `FlagManager` type of service that allows dynamic creation of flags and prevents same flag character conflict I opted to use hardcode character now for flag creation.

I am using `validate` now. With it the codebase is much leaner. However please not that that `validate` does not allow access to other flags. Furthermore, it is a static property in `flagsConfig` that is responsible for creating `SfdxFlag`s and therefore the `validate` function also needs to be static. This prevents the `validate` method to access or set instance properties. For sophisticated validation logic where flags are interdependent I think we will need to keep using something like `validateInputParameters` instead.

Reply:
Then in that case we should update `apiLevel` to `-l` instead of `-a` otherwise you'll have conflict in `Preview`.

Reply:
I started removing `CommandLineUtils.createFlagConfig` and instead was going to define the platform and api level flags in each of the command that use them. But I soon realized that if I do that the message resource also needs to be duplicated 😢 . So for now I'm going to keep `CommandLineUtils.createFlagConfig`. I changed the flag to be "-l" instead of "-a".